Improvements to animation keyframe computation

mrobinson opened the pull request:
This pull request contains two changes:

**Caching computed keyframes between animation changes**:
Instead of recomputing keyframe data for every tick of an animation,
cache the computed values when animations change. In addition to being
more efficient, this will allow us to return animation rules as property
declarations because we don't need to consult the final style to produce
them.

**Better computation of keyframe data**:
Instead of naively using `apply_declarations` to compute the style of
each keyframe, use an approach more like Gecko's where we carefully
walk through the animation keyframes and try to extract `AnimationValue`s
for each computed keyframe. In this approach we respect the order with
which the properties are set in the keyframe source and try to deal with
CSS custom properties.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
Cache animation computed values when animations change

Instead of recalculating the animation style every tick of an animation,
cache the computed values when animations change. In addition to being
more efficient, this will allow us to return animation rules as property
declarations because we don't need to consult the final style to produce
them.

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/style/properties/properties.mako.rs, components/style/animation.rs, components/style/matching.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit f4b736a7324d706343ff91e3e4268c2725ee1b87 with merge b6c8424937539a49af2b3a0665364dd1861f13e1...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/RyEwfcuNTxiC5twerao8hg)
State: approved= try=True

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26486) made this pull request unmergeable. Please resolve the merge conflicts.

mrobinson committed:
Better computation of animation keyframes

This begins to address #26625 by properly applying CSS variables during
keyframe computation and no longer using `apply_declarations`. Instead,
walk the declarations, combining them into IntermediateComputedKeyframe,
maintaining declarations that modify CSS custom properties. Then compute
a set of AnimationValues for each keyframe and use those to produce
interpolated animation values.

mrobinson commented:
I've added a closely related change to this PR which improves the computation of keyframe data and ditches Servo's use of `apply_declarations` for this.

mrobinson commented:
This needs a little bit more work.

jdm commented:
I read through 917350369e6251b4296d658f2ce388193d000603 and it looks sensible.

mrobinson commented:
@jdm Thanks for looking at the first commit. That should be unchanged in the newest version of the branch. The second commit is unfortunately a bit large, which means it's probably harder to review. I wasn't sure how to split it into simpler chunks. 😕 I'm happy to walk through it with anyone or do whatever else is necessary to make it easier to review.

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit b208d0421d767d39e6b203967761b7066efe807b with merge 643a386cb138ef6e0ff720c1bbd338c7d0e6593b...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Tbw1prerQEaM8PoVOanAaA)

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 4307b6324bc9327200cdc66ccedbe3a513c52235 with merge c3469c016e7c2c9db040e6f8976f15376875dc8b...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/CdHNSAFpQ7OkPUY_Cj1W6Q)
State: approved= try=True

mrobinson commented:
@bors-servo try=wpt-2020

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 4307b6324bc9327200cdc66ccedbe3a513c52235 with merge d22fd6c71e72d702445ddee4b6af8589c94ca7a4...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/fkG16769Tcqg7VPFCdnP9g)
State: approved= try=True

jdm reviewed (approved):
These changes look sensible to me.

emilio commented on components/style/animation.rs:
```diff
@@ -172,100 +173,223 @@ pub enum KeyframesIterationState {
+            self.timing_function = Some(timing_function.to_computed_value_without_context());
+        }
+
+        match step.value {
```
Maybe `let block = match setep.value { ... };`, then de-indent the rest?

mrobinson replied:
:+1:

emilio commented on components/style/animation.rs:
```diff
@@ -172,100 +173,223 @@ pub enum KeyframesIterationState {
+        &mut self,
+        step: &KeyframesStep,
+        context: &SharedStyleContext,
+        base_style: &Arc<ComputedValues>,
```
I think this can just be `&ComputedValues`, and same in the caller.

mrobinson replied:
Oh, nice.

emilio reviewed (approved):
Yeah, nice progression :)

mrobinson commented on components/style/animation.rs:
```diff
@@ -553,58 +677,46 @@ impl Animation {
-                from,
-                to,
+                from: from.clone(),
+                to: next_keyframe.values[index].clone(),
```
Your suggestion is a lot clearer.

mrobinson commented:
@bors-servo r=jdm,emilio

Thank you both for the reviews!

bors-servo commented:
:pushpin: Commit b875f14 has been approved by `jdm,emilio`

<!-- @bors-servo r=jdm,emilio b875f14e865084ddf596ed64455a022d386dcea7 -->

bors-servo commented:
:hourglass: Testing commit b875f14e865084ddf596ed64455a022d386dcea7 with merge ec49ebb753dcbc845a058cad853128ccdd31b3b0...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/R6itKLO4RnqHmI7He2FQQA)
Approved by: jdm,emilio
Pushing ec49ebb753dcbc845a058cad853128ccdd31b3b0 to master...
